File: Week12_LinkedList/reverse-list.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */

class Solution {
    public ListNode reverseList(ListNode head) {
        /**
         * If you do not understand this solution, please 
         * please use a piece of paper and marker to make
         * life easier for you!
         */

        /* Pointer to output Reversed Linked List */
        ListNode prev = null;
        /* Pointer to current Linked List */
        ListNode curr = head;

        while (curr != null) {
            /* Store temporary reference to next node */
            ListNode cacheNext = curr.next;
            /* Reversal of reference order occurs here */
            curr.next = prev;
            prev = curr;
            /* give curr back the reference to next node */
            curr = cacheNext;
        }

        return prev;
    }
}

File: Week12_LinkedList/reverse-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        """
        If you do not understand this solution, please
        please use a piece of paper and marker to make
        life easier for you!
        """
        # Pointer to output Linked List
        prev = None
        # Pointer to current Linked List
        curr = head
        
        while curr:
            cacheNext = curr.next
            curr.next = prev
            prev = curr
            curr = cacheNext
            
        return prev